### PR TITLE
PartitionedFileset Source and Sink Implementation with tests

### PR DESCRIPTION
**Pull request by rohitsinha54**
Implements: 
1. PartitionedFilesetSource: https://issues.cask.co/browse/CDAP-2063
2. PartitionedFilesetSink: https://issues.cask.co/browse/CDAP-2065
3. Unit test with PartitionedFilesetSource as source and PartitionedFilesetSink as a sink

This currently does not set explore enabled while creating the filesets. We need a small discussion on it and I will update this PR after the discussion.

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLPartitionedFileSetSourceSinkTest.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+    mapReduceArgs.put("config", GSON.toJson(adapterConfig));
+    MapReduceManager mrManager = batchManager.startMapReduce("ETLMapReduce", mapReduceArgs);
+    mrManager.waitForFinish(5, TimeUnit.MINUTES);
+    batchManager.stopAll();
+
+    verySinkData();
+  }
+
+  private void verySinkData() throws Exception {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLPartitionedFileSetSourceSinkTest.java]: typo: very -> verify

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLPartitionedFileSetSourceSinkTest.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    ApplicationManager batchManager = deployApplication(ETLBatchTemplate.class);
+
+    ApplicationTemplate<ETLBatchConfig> appTemplate = new ETLBatchTemplate();
+    ETLBatchConfig adapterConfig = constructETLBatchConfig();
+    MockAdapterConfigurer adapterConfigurer = new MockAdapterConfigurer();
+    appTemplate.configureAdapter("myAdapter", adapterConfig, adapterConfigurer);
+
+    Map<String, String> mapReduceArgs = Maps.newHashMap();
+    for (Map.Entry<String, String> entry : adapterConfigurer.getArguments().entrySet()) {
+      mapReduceArgs.put(entry.getKey(), entry.getValue());
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLPartitionedFileSetSourceSinkTest.java]: you can use the putAll method instead of iterating through all entries

Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLPartitionedFileSetSourceSinkTest.java]: you can also just pass the map into Maps.newHashMap

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLPartitionedFileSetSourceSinkTest.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+                                                     .addStringField("league", "nfl")
+                                                     .build());
+
+    Assert.assertNotNull("Partition not found", sinkPartition);
+    try {
+      Location location = sinkPartition.getLocation();
+      for (Location file : location.list()) {
+        if (file.getName().startsWith("part")) {
+          BufferedReader reader = new BufferedReader(new InputStreamReader(location.getInputStream()));
+          Assert.assertEquals("The data in source and sink is not same", reader.readLine(), DUMMY_RECORD);
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLPartitionedFileSetSourceSinkTest.java]: if surprised this passes.  Usually there are several empty part files along with the one with data.  Should make sure you're not reading empty data before checking this.

Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLPartitionedFileSetSourceSinkTest.java]: should close the reader too

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/PartitionedFileSetSink.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    if (!Strings.isNullOrEmpty(stageConfig.getProperties().get(PARTITIONING))) {
+      String pfsName = stageConfig.getProperties().get(FILESET_NAME);
+      String separator = stageConfig.getProperties().get(TextOutputFormat.SEPERATOR);
+      Preconditions.checkArgument(!Strings.isNullOrEmpty(pfsName), "PartitionedFileSet name must be given.");
+      Preconditions.checkArgument(!Strings.isNullOrEmpty(pfsName), "A separator for " +
+        TextOutputFormat.SEPERATOR + "must be given");
+      pipelineConfigurer.createDataset(pfsName, PartitionedFileSet.class.getName(),
+                                       PartitionedFileSetProperties.builder()
+        .setPartitioning(ETLUtils.createPartitioning(stageConfig.getProperties().get(PARTITIONING)))
+        .setInputFormat(TextInputFormat.class)
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/PartitionedFileSetSink.java]: as discussed, these should all be properties that the user sets when creating a pipeline

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/PartitionedFileSetSink.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    context.setOutput(context.getRuntimeArguments().get(FILESET_NAME), input);
+  }
+
+  /**
+   * Builds a {@link PartitionKey} from user supplied JSON string.
+   *
+   * @param paritionKeyString the partition key JSON string
+   * @return {@link PartitionKey}
+   */
+  private PartitionKey buildParitionKey(String paritionKeyString) {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/PartitionedFileSetSink.java]: typo: Parition -> Partition

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/PartitionedFileSetSink.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    PartitionKey.Builder builder = PartitionKey.builder();
+    Type stringStringMap = new TypeToken<Map<String, String>>() {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/PartitionedFileSetSink.java]: this type and the Gson instance can both be private static final variables

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/PartitionedFileSetSink.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+import java.lang.reflect.Type;
+import java.util.Map;
+
+/**
+ * CDAP {@link PartitionedFileSet} Batch Sink
+ */
+public class PartitionedFileSetSink extends BatchSink<Text, Text> {
+
+  private static final String FILESET_NAME = "filesetName";
+  private static final String PARTITION_KEY = "partitionKey";
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/PartitionedFileSetSink.java]: it is more accurate for this to be plural, "partitionKeys"

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/PartitionedFileSetSink.java (hunk 5 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/PartitionedFileSetSink.java]: also more accurate for this to be plural PartitionKeys

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/PartitionedFileSetSink.java [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    configurer.addProperty(new Property(PARTITIONING, "Partitioning to use for the Partitioned FileSet in JSON " +
+      "format. Should be provided if you want the pipeline to create the fileset. ex: " +
+      "{\"league\" : \"string\", \"season\" : \"int\"}", false));
+    configurer.addProperty(new Property(TextOutputFormat.SEPERATOR, "The output format separator. Must be provided " +
+      "if you want the pipeline to create the fileset", false));
+  }
+
+  @Override
+  public void configurePipeline(ETLStage stageConfig, PipelineConfigurer pipelineConfigurer) {
+    // if the partitioning is provided then we should try to create the fileset here
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/PartitionedFileSetSink.java]: should also check that required fields are not empty (like name and partition key)

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/PartitionedFileSetSource.java [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+      "Must be provided if you want the pipeline to create the fileset. ex: " +
+      "{\"league\" : \"string\", \"season\" : \"int\"}", false));
+    configurer.addProperty(new Property(TextOutputFormat.SEPERATOR, "The output format separator must be provided " +
+      "if you want the pipeline to create the fileset", false));
+
+  }
+
+  @Override
+  public void configurePipeline(ETLStage stageConfig, PipelineConfigurer pipelineConfigurer) {
+    // if the partitioning is provided then we should try to create the fileset here
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/PartitionedFileSetSource.java]: should also check that require fields like name are not missing

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/PartitionedFileSetSource.java [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+  /**
+   * creates a {@link PartitionFilter} from a given string in JSON format
+   *
+   * @param filterString partition filter JSON string
+   * @return {@link PartitionFilter}
+   */
+  private PartitionFilter createPartitionFilter(String filterString) {
+    PartitionFilter.Builder builder = PartitionFilter.builder();
+    Type stringStringMap = new TypeToken<Map<String, String>>() {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/PartitionedFileSetSource.java]: type and gson can be private static final variables

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/PartitionedFileSetSource.java [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+
+/**
+ * CDAP {@link PartitionedFileSet} Batch Source
+ */
+public class PartitionedFileSetSource extends BatchSource<LongWritable, Text> {
+
+  private static final String FILESET_NAME = "filesetName";
+  private static final String FILTER = "filter";
+  private static final String PARTITIONING = "partitioning";
+  private static final String VALUE = "value";
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/PartitionedFileSetSource.java]: should be '.value'

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/PartitionedFileSetSource.java [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+  private static final String LOWER = "lower";
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/PartitionedFileSetSource.java]: should be '.lower'

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/PartitionedFileSetSource.java [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+  private static final String UPPER = "upper";
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/PartitionedFileSetSource.java]: should be '.upper'

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/PartitionedFileSetSource.java [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+   */
+  private PartitionFilter createPartitionFilter(String filterString) {
+    PartitionFilter.Builder builder = PartitionFilter.builder();
+    Type stringStringMap = new TypeToken<Map<String, String>>() {
+    }.getType();
+    Map<String, String> filters = new Gson().fromJson(filterString, stringStringMap);
+    for (Map.Entry<String, String> entrySet : filters.entrySet()) {
+      String key = entrySet.getKey();
+      if (key.endsWith(VALUE)) {
+        builder.addValueCondition(key.split("\\.")[0], entrySet.getValue());
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/PartitionedFileSetSource.java]: if it is valid for a partition key to contain a '.', this will break.  It is better to find the last to take a substring from the beginning to the end - length of VALUE

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/PartitionedFileSetSource.java [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+      } else if (key.endsWith(LOWER)) {
+        String fieldName = key.split("\\.")[0];
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/PartitionedFileSetSource.java]: same comment as above

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/PartitionedFileSetSource.java [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+        // if the upper range is found for  this lowe range add it to PartitionFilter else just ignore
+        String upperRange = Joiner.on(".").join(fieldName, UPPER);
+        if (filters.containsKey(upperRange)) {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sources/PartitionedFileSetSource.java]: seems like we should error out if there is no upper range instead of silently ignoring it

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/common/ETLUtils.java [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+import com.google.common.base.Strings;
+import com.google.common.base.Throwables;
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+
+import java.lang.reflect.Type;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+/**
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/common/ETLUtils.java]: type and gson can be private static final instead of creating each time

File: cdap-integration-test/src/main/java/co/cask/cdap/test/IntegrationTestManager.java [hunk 18]
```diff
@@ -123,6 +124,12 @@ public void deployDatasetModule(Id.Namespace namespace, String moduleName,
   }
 
   @Override
+  public <T> DataSetManager<T> getDataset(Id.Namespace namespace, String datasetInstanceName,
```
Comment (albertshau) [inline cdap-integration-test/src/main/java/co/cask/cdap/test/IntegrationTestManager.java]: why is this needed?

Comment (albertshau) [inline cdap-integration-test/src/main/java/co/cask/cdap/test/IntegrationTestManager.java]: ok i see, added to superclass. ignore this comment.

**Comment by albertshau** [comment]
something is still missing from these... If you use the source or sink it will read the same data each time the workflow runs.  So for the sink, it will run once and then fail every time after that. And for the source, it will read the same data every time which does not seem useful.  In order for these to be useful, partitions must change in between runs.  The most common case I can think of is that there is some time based partition. What is the plan for this?

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLPartitionedFileSetSourceSinkTest.java [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+    Partition sinkPartition = sinkPFS.getPartition(PartitionKey.builder()
+                                                     .addStringField("league", "nfl")
+                                                     .build());
+
+    Assert.assertNotNull("Partition not found", sinkPartition);
+    try {
+      Location location = sinkPartition.getLocation();
+      for (Location file : location.list()) {
+        if (file.getName().startsWith("part")) {
+          BufferedReader reader = new BufferedReader(new InputStreamReader(location.getInputStream()));
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLPartitionedFileSetSourceSinkTest.java]: This is wrong. You always need a charset when creating `Reader` from `InputStream`. Also, better use:

`CharStreams.readFirstLine(CharStreams.newReaderSupplier(Locations.newInputSupplier(location), Charsets.UTF_8))`

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLPartitionedFileSetSourceSinkTest.java [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+
+    if (inputFileset.getPartition(key) != null) {
+      LOG.info("Partition already exists");
+      throw new RuntimeException("Partition already exists");
+    }
+    PartitionOutput output = inputFileset.getPartitionOutput(key);
+
+    try {
+      Location location = output.getLocation();
+      OutputStream outputStream = location.getOutputStream();
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLPartitionedFileSetSourceSinkTest.java]: The whole block can be replaced with one line:
```java
CharStreams.write(DUMMY_RECORD, CharStreams.newWriterSupplier(Locations.newOutputSupplier(location), Charsets.UTF_8));
```

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLPartitionedFileSetSourceSinkTest.java [hunk 21]
```diff
@@ -0,0 +1,176 @@
...
+      try {
+        outputStream.write(DUMMY_RECORD.getBytes(Charsets.UTF_8));
+      } finally {
+        outputStream.close();
+      }
+    } catch (IOException e) {
+      LOG.warn(String.format("Unable to write path '%s'", output.getRelativePath()));
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLPartitionedFileSetSourceSinkTest.java]: When logging an exception, always log with the exception as the last argument, otherwise there is no stacktrace for debug. Also, seems like you don't need to log and rethrow here. The junit show already be catching and logging and fail the test.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLPartitionedFileSetSourceSinkTest.java [hunk 22]
```diff
@@ -0,0 +1,176 @@
...
+          Assert.assertEquals("The data in source and sink is not same", reader.readLine(), DUMMY_RECORD);
+        }
+      }
+    } catch (IOException e) {
+      LOG.info("Unable to read path {}", sinkPartition.getRelativePath());
+    }
+  }
+
+  private void writeToSource() throws Exception {
+    DataSetManager<PartitionedFileSet> table1 = getDataset(sourceFileset);
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLPartitionedFileSetSourceSinkTest.java]: It is not a "table". Call the variable appropriately.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLPartitionedFileSetSourceSinkTest.java [hunk 23]
```diff
@@ -0,0 +1,176 @@
...
+      throw e;
+    }
+    output.addPartition();
+    table1.flush();
+  }
+
+  private ETLBatchConfig constructETLBatchConfig() {
+    ETLStage source = new ETLStage(PartitionedFileSetSource.class.getSimpleName(),
+                                   ImmutableMap.of("filesetName", sourceFileset, "filter",
+                                                   "{\"league.value\": \"nfl\",\"season.lower\": " +
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLPartitionedFileSetSourceSinkTest.java]: Instead of hand writing Json, why not use Gson?

```java
GSON.toJson(ImmutableMap.of("league.value", "nfl", "season.lower", "1980", "season.upper", "1990"))
```

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLStreamConversionTest.java [hunk 24]
```diff
@@ -0,0 +1,161 @@
...
+    transformList.add(streamToStructuredRecord);
+    transformList.add(structuredRecordToGeneric);
+    transformList.add(genericToAvro);
+    return new ETLBatchConfig("", source, sink, transformList);
+  }
+
+  private List<GenericRecord> readOutput(TimePartitionedFileSet fileSet, Schema schema) throws IOException {
+    org.apache.avro.Schema avroSchema = new org.apache.avro.Schema.Parser().parse(schema.toString());
+    DatumReader<GenericRecord> datumReader = new GenericDatumReader<GenericRecord>(avroSchema);
+    List<GenericRecord> records = com.google.common.collect.Lists.newArrayList();
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLStreamConversionTest.java]: why full name? You have other `Lists`?

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLStreamConversionTest.java [hunk 25]
```diff
@@ -0,0 +1,161 @@
...
+      // this level should be the day (ex: 2015-01-19)
+      for (Location timeLoc : dayLoc.list()) {
+        // this level should be the time (ex: 21-23.1234567890000)
+        for (Location file : timeLoc.list()) {
+          // this level should be the actual mapred output
+          String locName = file.getName();
+
+          if (locName.endsWith(".avro")) {
+            DataFileStream<GenericRecord> fileStream =
+              new DataFileStream<GenericRecord>(file.getInputStream(), datumReader);
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLStreamConversionTest.java]: Close stream after used in finally block.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLStreamConversionTest.java [hunk 26]
```diff
@@ -0,0 +1,161 @@
...
+            while (fileStream.hasNext()) {
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/templates/etl/batch/ETLStreamConversionTest.java]: Just use `Iterables.addAll(records, fileStream)`.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/PartitionedFileSetSink.java [hunk 27]
```diff
@@ -0,0 +1,110 @@
...
+  private static final String PARTITION_KEY = "partitionKey";
+  private static final String PARTITIONING = "partitioning";
+
+  @Override
+  public void configure(StageConfigurer configurer) {
+    configurer.setName(PartitionedFileSetSink.class.getSimpleName());
+    configurer.setDescription("CDAP Partitioned fileSet batch sink");
+    configurer.addProperty(new Property(FILESET_NAME, "Partitioned fileset name", true));
+    configurer.addProperty(new Property(PARTITION_KEY, "Partition key in JSON format to use while writing " +
+      "ex: {\"league\" : \"nfl\", \"season\" : \"1980\"}", true));
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/batch/sinks/PartitionedFileSetSink.java]: Same comment as above. Use GSON to generate json string. don't hand write it.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/common/ETLUtils.java [hunk 28]
```diff
@@ -0,0 +1,89 @@
...
+ * An Utility class for ETL
+ */
+public class ETLUtils {
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/common/ETLUtils.java]: Util class must be `final` with a private constructor.

Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/common/ETLUtils.java]: Also, I don't like a class call `XXXUtils`, as it means you can put anything there, which is super bad (like a random collection of stuff, which is a signal for bad OO design sometime).

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/common/ETLUtils.java [hunk 29]
```diff
@@ -0,0 +1,89 @@
...
+    //TODO: replace with TimeMathParser (available only internal to cdap)
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(frequencyStr));
+    frequencyStr = frequencyStr.toLowerCase();
+
+    String value = frequencyStr.substring(0, frequencyStr.length() - 1);
+    int parsedValue = 0;
+    try {
+      parsedValue = Integer.parseInt(value);
+    } catch (NumberFormatException nfe) {
+      Throwables.propagate(nfe);
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/common/ETLUtils.java]: Better write `throw Throwables.propagate(nfe)`.

**Comment by rohitsinha54** [comment]
@albertshau : For a time based partition we have the TimeParitionedFilset and the sink for the same. Is it not something which is more appropriate for it?  If we take a step back and look at it we had the same issue with FileSets as there was no way in the current ETL architecture to write to a new inputPath file in every run as the parameters are passed just once. I brought this up in the Adapter room and we made a consensus that FileSet  are very low level and will not be so much useful which made us switch to PartitionedFileSet: 

```
[Apr-9 12:42 PM] Rohit Sinha: @terence : Not we are using FileSets to make a FileSetSource.
[Apr-9 12:47 PM] Rohit Sinha: No*
[Apr-9 12:52 PM] Terence Yim: Why not use the ParititonFileSet?
[Apr-9 12:52 PM] Invisible Joltie: +!
[Apr-9 12:52 PM] Terence Yim: or that should be a different sink?
[Apr-9 12:52 PM] Terence Yim: the raw FileSet is very limited in feature.
[Apr-9 12:52 PM] Terence Yim: and PartitionedFileSet should be used in most cases.
[Apr-9 12:53 PM] Terence Yim: The FileSet is mainly for lower level usage.. e.g. you already have some existing file on HDFS and wanted to use them in CDAP... or for being used inside a PartitionedFileSet.
[Apr-9 1:31 PM] Rohit Sinha: The requirement was to create FileSetSource, FileSetSink and PartitionedFileSetSink. I am not sure as how we came up with these.
[Apr-9 1:31 PM] Rohit Sinha: @sree ^^
[Apr-9 1:31 PM] Rohit Sinha: Should we just do Partitioned then ?
[Apr-9 1:36 PM] Sreevatsan Raman: the requirement is to integrate with Datasets. If we think FileSets doesn't make sense then let's do paritioned file sets
[Apr-9 2:01 PM] Terence Yim: Remember, the goal is to create something useful...
[Apr-9 2:02 PM] Terence Yim: not just to create a collection of semi-useful/not-useful stuff for our users.
```

More or less with this too we are at same issue. The use case which I see here with PartitionFilesetSource and Sink is to have Workflow triggered by new partition being created.

**Comment by rohitsinha54** [comment]
So after discussion with @sreevatsanraman and @albertshau we decided to stay put on the PFS source and sink tell we fix the underlying architecture which will enable us to write to existing partition when new data comes in through MR or if ETL can support changing arguments in consecutive runs.
We will be closing this PR for now and will come back to it when we have architecture changes
